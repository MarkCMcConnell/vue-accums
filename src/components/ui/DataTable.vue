<template>
    <table class="table table-striped">
        <thead>
            <tr>
                <th
                scope="col"
                @click.prevent="sortColumn(field.field)"
                v-for="field in fields"
                :key="field.label"
                >{{ field.label }}</th>
            </tr>
        </thead>
        <tbody>
            <tr v-for="(obj, index) in sortedData" :key="index">
                <td v-for="field in fields" :key="field.field">
                <span v-if="field.field == 'ProgressBar'">
                    <ProgressBar :width="obj[field.field]"></ProgressBar>
                </span>
                <span
                    v-else-if="field.field == 'ClaimSystemCode'"
                    :class="{ 'gapSoon': obj[field.field] === 'Due Soon', 'gapRed': obj[field.field] === 'Overdue', 'gapNew': obj[field.field] === 'New' }"
                >{{ obj[field.field]}}</span>
                <span v-else>{{ obj[field.field] }}</span>
                </td>
            </tr>
        </tbody>
    </table>
</template>

<script>
import ProgressBar from './ProgressBar.vue'

export default {
    components: {
        ProgressBar
    },
    props: {
        fields: Array,
        data: Array
    },
    data () {
        return {
            currentSort: this.fields[0].label,
            currentDirection: 'ascending',
            tableData: this.data
        }
    },
    computed: {
        sortedData: function () {
            const sortedData = this.tableData

            return sortedData.sort((a, b) => {
                let modifier = 1

                if (this.currentDirection === 'descending') {
                    modifier = -1
                }

                if (a[this.currentSort] < b[this.currentSort]) {
                    return -1 * modifier
                } else if (a[this.currentSort] > b[this.currentSort]) {
                    return 1 * modifier
                } else {
                    return 0
                }
            })
        }
    },
    methods: {
        sortColumn (column) {
            if (column === this.currentSort) {
                this.currentDirection =
                    this.currentDirection === 'ascending' ? 'descending' : 'ascending'
            }

            this.currentSort = column
        }
    }
}
</script>

<style scoped lang="scss">
.gapSoon {
    color: #c7b523;
}

.gapRed {
    color: #f00;
}

.gapNew {
    color: #2f9805;
}

th {
    cursor: pointer;
    text-decoration: underline;
    color: red;
}
</style>
