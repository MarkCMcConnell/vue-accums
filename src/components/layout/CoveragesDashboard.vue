<template>
  <section
    :class="this.isDashboardFormat ? 'dashboard' : ''"
  >
  <!-- TODO: Slot or props for title -->
    <h3>Coverages</h3>
    <DataTable
      :data="formatCoverages"
      :fields="coverageFields"
    />
  </section>
</template>

<script>
// General UI Components
import DataTable from '../ui/DataTable.vue'
// Mixins
import { dataHelpers } from '../../mixins/dataHelpers.js'

export default {
  components: {
    DataTable
  },
  props: {
    coverageData: {
      type: Array,
      required: true
    },
    isDashboardFormat: {
      type: Boolean
    },
    yearRange: {
      type: Number
    }
  },
  mixins: [dataHelpers],
  data () {
    return {
      coverageFields: [
        {
          label: 'Name',
          field: 'CoverageName'
        },
        {
          label: 'Type',
          field: 'CovPlan'
        },
        {
          label: 'Effective Date',
          field: 'EffectiveDate'
        },
        {
          label: 'Term Date',
          field: 'TermDate'
        }
      ]
    }
  },
  computed: {
    formatCoverages () {
      let coverages = this.coverageData
      let dataValues = []

      if (this.yearRange) {
        coverages = coverages.filter(coverage => {
          const currentYear = new Date().getFullYear()
          let date = this.convertJSONDateToDateString(coverage.EffectiveDate)
          
          return parseInt(date.substring(date.length - 4), 10) >= (currentYear - this.yearRange)
        })
      }

      for (let obj of coverages) {
        let data = {}

        for (let value of this.coverageFields) {
          let currentKey = obj[value.field]
          // Capture dates and convert /Date()/ format to human readable
          if (currentKey.substring(0, 5) === '/Date') {
            currentKey = this.convertJSONDateToDateString(currentKey)
          }
          // Avoid setting an empty row
          if (currentKey !== '' && currentKey !== undefined) {
            data[value.field] = currentKey
          }
        }
        // Push to the array that will be passed to the coverages table
        dataValues.push(data)
      }
      return dataValues
    }
  }
}
</script>

<style lang="scss">
  .dashboard {
    margin-top: 2rem;
    padding: 1.25rem;
    width: 60%;
    box-shadow: 0 10px 20px rgba(0, 0, 0, 0.19), 0 6px 6px rgba(0, 0, 0, 0.23);
  }
</style>
