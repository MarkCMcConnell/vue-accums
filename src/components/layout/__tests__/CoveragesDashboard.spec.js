/* eslint-disable */
import { mount } from '@vue/test-utils'

import CoveragesDashboard from '../CoveragesDashboard.vue'
import DataTable from '../../ui/DataTable.vue'

describe('CoveragesDashboard', () => {
	let wrapper

	beforeEach(() => {
		wrapper = mount(CoveragesDashboard, {
			propsData: {
				coverageData: [
					{
						CoverageName: 'DENTAL',
						CovPlan: 'Silver',
						EffectiveDate: '\/Date(1514782800000-0500)\/',
						TermDate: '\/Date(253402232400000-0500)\/'
					},
					{
						CoverageName: 'DENTAL',
						CovPlan: 'Silver',
						EffectiveDate: '\/Date(1483246800000-0500)\/',
						TermDate: '\/Date(1514696400000-0500)\/'
					}
				],
				isDashboardFormat: true,
				yearRange: 1
			}
		})
	})

	afterEach(() => {
		wrapper.destroy()
	})

	it('renders the component', () => {
		expect(wrapper.html()).toMatchSnapshot()
	})

	describe('dashboard format', () => {
		it('renders as a dashboard based on isDashboardFormat prop', () => {
			const dashboardWrapper = wrapper.find('section')
			expect(dashboardWrapper.classes()).toContain('dashboard')
		})
	})

	describe('formatCoverages', () => {
		it('converts Healthx Date string to human readable', () => {
			expect(wrapper.vm.formatCoverages).toEqual(
				expect.arrayContaining([
					expect.objectContaining({
						EffectiveDate: '1/1/2018'
					})
				])
			)
		})

		it('renders the specified year range', () => {
			expect(wrapper.findAll('.dataRow').length).toBe(1)
		})
	})
})