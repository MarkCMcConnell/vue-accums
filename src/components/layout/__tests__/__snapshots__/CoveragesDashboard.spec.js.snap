// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoveragesDashboard renders the component 1`] = `
<section class="dashboard">
  <h3>Coverages</h3>
  <table class="table table-striped">
    <thead>
      <tr>
        <th scope="col">Name</th>
        <th scope="col">Type</th>
        <th scope="col">Effective Date</th>
        <th scope="col">Term Date</th>
      </tr>
    </thead>
    <tbody>
      <tr class="dataRow">
        <td><span>DENTAL</span></td>
        <td><span>Silver</span></td>
        <td><span>1/1/2018</span></td>
        <td><span>12/31/9999</span></td>
      </tr>
    </tbody>
  </table>
</section>
`;
